src/pybind/mgr/progress/module.py

```diff
@@ -607,53 +620,59 @@ def _pg_state_changed(self, pg_dump):
         except ZeroDivisionError:
             return
         if progress < 1.0:
+            self.log.warning("Starting Global Recovery Event,\
+              {0} pgs not in active + clean state \
+              ".format(total_pg_num - active_clean_num))
             ev = GlobalRecoveryEvent("Global Recovery Event",
-                    refs=[("global","")],
+                    refs=[("global", "")],
                     add_to_ceph_s=True,
                     start_epoch=self.get_osdmap().get_epoch(),
                     active_clean_num=active_clean_num)
-            ev.global_event_update_progress(pg_dump)
+            ev.global_event_update_progress(self.get('pg_stats'), self.log)
             self._events[ev.id] = ev
 
-    def notify(self, notify_type, notify_data):
-        self._ready.wait()
+    def _process_new_osd_map(self):
+        old_osdmap = self._latest_osdmap
+        self._latest_osdmap = self.get_osdmap()
+        assert old_osdmap
+        assert self._latest_osdmap
+
+        self.log.info("Processing OSDMap change {0}..{1}".format(
+            old_osdmap.get_epoch(), self._latest_osdmap.get_epoch()
+        ))
+        self._osdmap_changed(old_osdmap, self._latest_osdmap)
+
+    def _process_new_pg_map(self):
+        # if there are no events we will skip this here to avoid
+        # expensive get calls
+        if len(self._events) == 0:
+            return
+
+        global_event = False
+        data = self.get("pg_stats")
+        ready = self.get("pg_ready")
+        for ev_id in list(self._events):
+            ev = self._events[ev_id]
```

Comment:
if what we care are but the values of the dict. why not put
```py
for ev in self._events.values():
  # ...
```

Reply:
same reason as previous comment about runtime error when items in dictionary are changed